README
libtextwrap - Text-Wrapping Library with I18N


Introduction
------------

Word-wrapping (or line-folding) is one of frequently used text-processing.
It is implemented very easily only when it is based on several assumptions
such as:

 - One character consists from one byte.
 - One character occupies one column on the terminal.
 - A line can be folded between words which are separated by whitespaces.

All of these assumptions are *local* and valid only for some
part of languages in the world.  There are languages which don't meet
the assumptions and speakers of such languages cannot use softwares
which depend on these assumptions.

Since it is usually difficult to require developers to know various
languages in the world and their properties, text-processing
algorithms should be supplied as libraries.  Otherwise, their softwares
cannot support various languages in the world and people in the world
suffer from a problem that many softwares are useless for them.


Feature
-------

On calling a function with parameters of text and maximum columns,
one can get line-folded text.  The text is encoded in current LC_CTYPE
locale.  This means that libtextwrap uses UTF-8 for I/O in UTF-8 locales,
ISO-8859-1 in ISO-8859-1 locales, EUC-JP in EUC-JP locales, KOI8-R in
KOI8-R locales, and so on.


Download
--------

Please visit http://libtextwrap.sourceforge.net/ .


Compilation
-----------

When you are using released version, do like following:

$ ./configure
$ make
$ make install

When you are using CVS version, you will have to run

$ ./bootstrap

before invoking ./configure script.  You will need autoconf,
automake, and libtool in this case.


Usage
-----

See the manpage of textwrap(3).  Like following:

#include <stdio.h>
#include <locale.h>
#include <textwrap.h>
int main(int argc, char **argv)
{
    textwrap_t property;

    setlocale(LC_ALL, "");
    textwrap_init(&property);
    textwrap_columns(&property, 40);

    printf("Input string:\n%s\n", argv[1]);
    printf("Formatted string:\n%s\n",
        textwrap(&property, argv[1]));
}


Sample Program
--------------

dotextwrap.c is a test program to invoke libtextwrap.
It is compiled and installed at the same time as libtextwrap.
Read dotextwrap(1) for detail.

sample/testprog.pl is a test program for libtextwrap with
multilingual text.  Just invoke it.


License
-------

Copyright (c) 2003 by Tomohiro KUBOTA <kubota@debian.org>

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
3. The name of any author may not be used to endorse or promote
   products derived from this software without their specific prior
   written permission.

THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.



The sample texts in sample/testprog.pl are citations from Debian website
(http://www.debian.org/).

Copyright (c) 1997-2003 Software in the Public Interest, Inc.
P.O. Box 502761
Indianapolis, IN 46250-2761
United States

This material may be distributed only subject to the terms and conditions
set forth in the Open Publication License, Draft v1.0 or later (the latest
version is presently available at http://www.opencontent.org/openpub/). 

"Debian" and the Debian Logo are trademarks of Software in the Public
Interest, Inc.


Changelog
2003-08-24  Tomohiro KUBOTA <kubota@debian.org>

	* API was changed for consistent handling of textwrap_t.
	* Tab code support.
	* Added a small program "dotextwrap" and its manpage.
	* Manpage improvement.
	* sample/testprog.c is obsoleted by dotextwrap.
	* version 0.1

2003-08-23  Tomohiro KUBOTA <kubota@debian.org>

	* Added "libtoolize" to bootstrap so that it works well
	  with automake 1.6 or later.
	* The manpage is improved.

2003-08-21  Tomohiro KUBOTA <kubota@debian.org>

	* API was changed to use textwrap_t.
	* Added a small sample program.

2003-08-19  Tomohiro KUBOTA <kubota@debian.org>

	* README was improved.  Instruction for compilation and
	  usage was added.

2003-08-16  Tomohiro KUBOTA <kubota@debian.org>

	* version 0.01
	* Initial version.


textwrap.c
/*
 * libtextwrap - Text-Wrapping Library with I18N
 * Copyright (C) 2003 by Tomohiro KUBOTA <kubota@debian.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of any author may not be used to endorse or promote
 *    products derived from this software without their specific prior
 *    written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <langinfo.h>
#include <ctype.h>
#include <wchar.h>

#include "textwrap.h"

typedef struct {char *buf; int size, len;} string_t;

#define ENCODING_TYPE_SIMPLE	0
#define ENCODING_TYPE_UTF8	1
#define ENCODING_TYPE_CJ	2

/* ---------- internal functions ---------- */

static string_t *
stringt_new(int size)
{
  string_t *p;
  p = (string_t *)malloc(sizeof(string_t)); if (!p) return NULL;
  p->buf = (char *)malloc(size); if (!p->buf) { free(p); return NULL; }
  p->size = size;
  p->len = 0;
  p->buf[0] = 0;
  return p;
}

static void
stringt_destroy(string_t *p)
{
  free(p->buf);
  free(p);
  return;
}

static char *
stringt_destroy_extract(string_t *p)
{
  char *s = p->buf;
  free(p);
  return s;
}

static char *
stringt_extract(string_t *p)
{
  return p->buf;
}

static void
stringt_zero(string_t *p)
{
  p->len = 0;
  p->buf[0] = 0;
}

static string_t *
stringt_addstrn(string_t *p, const char *c, int len)
{
  if (c == NULL) return p;
  if (p->len + len + 1 > p->size) {
    p->buf = (char *)realloc(p->buf, (p->size + len + 1) * 2);
    if (!p->buf) { free(p); return NULL;}
    p->size = (p->size + len + 1) * 2;
  }
  strncpy(p->buf + p->len, c, len);
  p->buf[p->len + len] = 0; p->len += len;
  return p;
}

static string_t *
stringt_addstr(string_t *p, const char *c)
{
  if (!c) return p;
  return stringt_addstrn(p, c, strlen(c));
}

static string_t *
stringt_addstringt(string_t *p, string_t *q)
{
  stringt_addstrn(p, q->buf, q->len);
}

/*
 * Since the return value of nl_langinfo() may differ depending on
 * systems (such as "EUC-JP" vs "eucJP"), the encoding names are
 * normalized (all-uppercase, eliminate non-alphanumerics).
 */

static char *
normalize(char *str)
{
  char *out, *p, *q;

  p = str;
  q = out = (char *)malloc(strlen(str) + 1);

  while(*p) {
    *q = toupper(*p++);
    if (isalnum(*q)) q++;
  }
  *q = 0;
  return out;
}

/*
 * "Encoding type" affects breakable().  It doesn't affect number of
 * bytes of one character nor number of columns of one character.
 */

int
set_encoding_type(const char *encoding)
{
  if (!strcmp(encoding, "UTF8")) return ENCODING_TYPE_UTF8;
  else if (strstr(encoding, "EUCKR")) return ENCODING_TYPE_SIMPLE;
  else if (strstr(encoding, "BIG5")) return ENCODING_TYPE_CJ;
  else if (strstr(encoding, "EUC")) return ENCODING_TYPE_CJ;
  else if (strstr(encoding, "GB2312")) return ENCODING_TYPE_CJ;
  return ENCODING_TYPE_SIMPLE;
}

/*
 * "multibyte character" version of wcwidth(3).
 */
static int
mbwidth(const char *ch, int ml)
{
  wchar_t c[2];

  /* 0x00 - 0x7f is always ASCII */
  if ((*ch >= 0 && *ch < 32) || *ch == 127) return -1;
  if (*ch >= 0 && *ch < 127) return 1;
  if (!mbtowc(c, ch, ml)) return -1;
  return wcwidth(c[0]);
}

/*
 * returns 1 if the character allows line-breaking after it.
 * the character is given in "multibyte character" in C meaning,
 * i.e., the encoding specified by LC_CTYPE locale.
 */

static int
breakable(const char *ch, int ml, int encoding_type)
{
  unsigned int u;

  if (*ch == ' ' || *ch == '\t') return 1;
  switch(encoding_type){
  case ENCODING_TYPE_SIMPLE: return 0;
  case ENCODING_TYPE_CJ: return (ml >= 2);
  case ENCODING_TYPE_UTF8:
    switch (ml){
    case 3: /* U+0800 - U+FFFF */
      u = (*ch&0x0f)*0x1000 + (*(ch+1)&0x3f)*0x40 + (*(ch+2)&0x3f);
      if (u >= 0x3000 && u <= 0x312f) {
	if (u == 0x300a || u == 0x300c || u == 0x300e || u == 0x3010
	    || u == 0x3014 || u == 0x3016 || u == 0x3018 || u == 0x301a)
	  return 0;
	return 1;
      }  /* CJK punctuations, Hiragana, Katakana, Bopomofo */
      if (u >= 0x31a0 && u <= 0x31bf) return 1;  /* Bopomofo */
      if (u >= 0x31f0 && u <= 0x31ff) return 1;  /* Katakana extension */
      if (u >= 0x3400 && u <= 0x9fff) return 1;  /* Han Ideogram */
      if (u >= 0xf900 && u <= 0xfaff) return 1;  /* Han Ideogram */
      return 0;
    case 4: /* U+10000 - U+1FFFFF */
      u = (*ch&7)*0x40000 + (*(ch+1)&0x3f)*0x1000 + (*(ch+2)&0x3f)*0x40
	+ (*(ch+3)&0x3f);
      if (u >= 0x20000 && u <= 0x2ffff) return 1;  /* Han Ideogram */
      return 0;
    default: return 0;
    }
  }    
}

/* ---------- public functions ---------- */

void
textwrap_init(textwrap_t *prop)
{
  prop->columns = 76;
  prop->tab = 8;
  prop->head1 = NULL;
  prop->head2 = NULL;
}

void
textwrap_columns(textwrap_t *prop, int columns)
{
  if (columns > 0) prop->columns = columns;
}

void
textwrap_tab(textwrap_t *prop, int tab)
{
  if (tab > 0) prop->tab = tab;
}

void
textwrap_head(textwrap_t *prop, const char *head1, const char *head2)
{
  prop->head1 = head1;
  prop->head2 = head2;
}

char *
textwrap(const textwrap_t *prop, const char *text)
{
  int len;
  string_t *out, *word;
  int additional;
  const char *p;
  int line_width, word_width;
  int head1_width, head2_width;
  char *encoding;
  int encoding_type;
  int columns, tab;
  const char *head1, *head2;

  columns = prop->columns; tab = prop->tab;
  head1 = prop->head1; head2 = prop->head2;

  /* locale initialization */
  encoding = normalize(nl_langinfo(CODESET));
  encoding_type = set_encoding_type(encoding);

  /* buffer initialization */
  len = strlen(text);
  additional = len / columns + columns;
  out = stringt_new(len + additional);
  word = stringt_new(len);

  /* header initialization */
  if (head1 != NULL) {
    for (p = head1, head1_width = 0; *p;) {
      int l = mblen(p, MB_CUR_MAX);
      if (l < 0) {head1 = NULL; head1_width = 0; break;}
      head1_width += mbwidth(p, l); p += l;
    }
    if (head1_width >= columns) {head1 = NULL; head1_width = 0;}
  } else head1_width = 0;

  if (head2 != NULL) {
    for (p = head2, head2_width = 0; *p;) {
      int l = mblen(p, MB_CUR_MAX);
      if (l < 0) {head2 = NULL; head2_width = 0; break;}
      head2_width += mbwidth(p, l); p += l;
    }
    if (head2_width >= columns) {head2 = NULL; head2_width = 0;}
  } else head2_width = 0;

  /* preparation for the main loop */
  p = text; /* next input character */
  stringt_addstr(out, head1);
  line_width = head1_width;
  word_width = 0;

  /* main loop */
  while(1){
    int ml, w, b;
    const char *now;

    if (!*p) {
      stringt_addstringt(out, word);
      return stringt_destroy_extract(out);
    }

    now = p; /* current character */
    ml = mblen(p, MB_CUR_MAX);
    w = mbwidth(p, ml);
    b = breakable(p, ml, encoding_type);
    p += ml;

    /*
     * newline code
     */

    if (*now == '\n') {
      stringt_addstringt(out, word);
      stringt_addstr(out, "\n");
      if (!*p) return stringt_destroy_extract(out);
      stringt_addstr(out, head2); line_width = head2_width;
      stringt_zero(word); word_width = 0;
      continue;
    }

    /*
     * tab code
     */

    if (*now == '\t') {
      if (line_width + word_width >= (columns / tab) * tab) {
	/* new line */
	stringt_addstringt(out, word);
	stringt_addstr(out, "\n");
	if (!*p) return stringt_destroy_extract(out);
	stringt_addstr(out, head2); line_width = head2_width;
	stringt_zero(word); word_width = 0;
      } else {
	/* not new line */
	int j;
	stringt_addstringt(out, word); line_width += word_width;
	stringt_zero(word); word_width = 0;
	j = tab - (line_width + word_width) % tab;
	for( ; j; j--) { stringt_addstr(out, " "); line_width ++; }
      }
      continue;
    }

    if (w < 0 || (*now >= 0 && *now < 0x20)) continue;

    /*
     * when the current line have enough room for the current character
     */

    if (line_width + word_width + w <= columns) {
      if (*now == ' ' || b) {
	stringt_addstringt(out, word);
	stringt_addstrn(out, now, ml);
	line_width += word_width + w;
	stringt_zero(word); word_width = 0;
      } else {
	stringt_addstrn(word, now, ml); word_width += w;
      }
      continue;
    }

    /*
     * when the current line overflows with the current character
     */

    if (*now == ' ') {
      stringt_addstringt(out, word);
      stringt_addstr(out, "\n");
      stringt_addstr(out, head2); line_width = head2_width;
      stringt_zero(word); word_width = 0;
    } else if (word_width + w <= columns) {
      stringt_addstr(out, "\n");
      stringt_addstr(out, head2); line_width = head2_width;
      p = now;
    } else {
      stringt_addstringt(out, word);
      stringt_addstr(out, "\n");
      stringt_addstr(out, head2); line_width = head2_width;
      stringt_zero(word);
      stringt_addstrn(word, now, ml); word_width = w;
    }
  }
}


textwrap.h
/*
 * libtextwrap - Text-Wrapping Library with I18N
 * Copyright (C) 2003 by Tomohiro KUBOTA <kubota@debian.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of any author may not be used to endorse or promote
 *    products derived from this software without their specific prior
 *    written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#ifndef __TEXTWRAP_H__
#define __TEXTWRAP_H__


typedef struct {
  int columns;
  int tab;
  const char *head1;
  const char *head2;
} textwrap_t;


#ifdef __cplusplus
extern "C"
{
#endif /* __cplusplus */

void textwrap_init(textwrap_t *prop);
void textwrap_columns(textwrap_t *prop, int columns);
void textwrap_tab(textwrap_t *prop, int tab);
void textwrap_head(textwrap_t *prop, const char *head1, const char *head2);

char *textwrap(const textwrap_t *prop, const char *text);

#ifdef __cplusplus
}
#endif /* __cplusplus */

#endif /* __TEXTWRAP_H__ */


sample/testprog.pl
#!/usr/bin/perl

# libtextwrap - Text-Wrapping Library with I18N
# Copyright (C) 2003 by Tomohiro KUBOTA <kubota@debian.org>
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. The name of any author may not be used to endorse or promote
#    products derived from this software without their specific prior
#    written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.


# The sample texts are citations from Debian website (http://www.debian.org/).
#
# Copyright (c) 1997-2003 Software in the Public Interest, Inc.
# P.O. Box 502761
# Indianapolis, IN 46250-2761
# United States
#
# This material may be distributed only subject to the terms and conditions
# set forth in the Open Publication License, Draft v1.0 or later (the latest
# version is presently available at http://www.opencontent.org/openpub/). 
#
# "Debian" and the Debian Logo are trademarks of Software in the Public
# Interest, Inc.


$locale = `locale charmap`;
chomp($locale);

print "locale = $locale\n";

if ($locale eq "UTF-8") {
    $str = <<"EOF";
English - Debian is a free operating system (OS) for your computer. An operating system is the set of basic programs and utilities that make your computer run.
Greek - \xce\xa4\xce\xbf Debian \xce\xb5\xce\xaf\xce\xbd\xce\xb1\xce\xb9 \xce\xad\xce\xbd\xce\xb1 \xce\xb5\xce\xbb\xce\xb5\xcf\x8d\xce\xb8\xce\xb5\xcf\x81\xce\xbf \xce\xbb\xce\xb5\xce\xb9\xcf\x84\xce\xbf\xcf\x85\xcf\x81\xce\xb3\xce\xb9\xce\xba\xcf\x8c \xcf\x83\xcf\x8d\xcf\x83\xcf\x84\xce\xb7\xce\xbc\xce\xb1 (\xce\x9b\xce\xa3) \xce\xb3\xce\xb9\xce\xb1 \xcf\x84\xce\xbf\xce\xbd \xcf\x85\xcf\x80\xce\xbf\xce\xbb\xce\xbf\xce\xb3\xce\xb9\xcf\x83\xcf\x84\xce\xae \xcf\x83\xce\xb1\xcf\x82. \xce\xa4\xce\xbf \xce\xbb\xce\xb5\xce\xb9\xcf\x84\xce\xbf\xcf\x85\xcf\x81\xce\xb3\xce\xb9\xce\xba\xcf\x8c \xcf\x83\xcf\x8d\xcf\x83\xcf\x84\xce\xb7\xce\xbc\xce\xb1 \xce\xb5\xce\xaf\xce\xbd\xce\xb1\xce\xb9 \xce\xbc\xce\xb9\xce\xb1 \xcf\x83\xcf\x85\xce\xbb\xce\xbb\xce\xbf\xce\xb3\xce\xae \xce\xb1\xcf\x80\xce\xbf \xcf\x84\xce\xb1 \xce\xb1\xcf\x80\xce\xb1\xcf\x81\xce\xb1\xce\xaf\xcf\x84\xce\xb7\xcf\x84\xce\xb1 \xcf\x80\xcf\x81\xce\xbf\xce\xb3\xcf\x81\xce\xac\xce\xbc\xce\xbc\xce\xb1\xcf\x84\xce\xb1 \xce\xba\xce\xb1\xce\xb9 \xce\xb2\xce\xbf\xce\xb7\xce\xb8\xce\xae\xce\xbc\xce\xb1\xcf\x84\xce\xb1 \xcf\x80\xce\xbf\xcf\x85 \xce\xb1\xcf\x80\xce\xb1\xce\xb9\xcf\x84\xce\xb5\xce\xaf \xce\xbf \xcf\x85\xcf\x80\xce\xbf\xce\xbb\xce\xbf\xce\xb3\xce\xb9\xcf\x83\xcf\x84\xce\xae\xcf\x82 \xcf\x83\xce\xb1\xcf\x82 \xce\xb3\xce\xb9\xce\xb1 \xce\xbd\xce\xb1 \xce\xbb\xce\xb5\xce\xb9\xcf\x84\xce\xbf\xcf\x85\xcf\x81\xce\xb3\xce\xae\xcf\x83\xce\xb5\xce\xb9.
Japanese - Debian \xe3\x81\xaf\xe3\x81\x82\xe3\x81\xaa\xe3\x81\x9f\xe3\x81\xae\xe3\x82\xb3\xe3\x83\xb3\xe3\x83\x94\xe3\x83\xa5\xe3\x83\xbc\xe3\x82\xbf\xe3\x81\xae\xe3\x81\x9f\xe3\x82\x81\xe3\x81\xae\xe3\x83\x95\xe3\x83\xaa\xe3\x83\xbc\xe3\x81\xaa\xe3\x82\xaa\xe3\x83\x9a\xe3\x83\xac\xe3\x83\xbc\xe3\x83\x86\xe3\x82\xa3\xe3\x83\xb3\xe3\x82\xb0\xe3\x82\xb7\xe3\x82\xb9\xe3\x83\x86\xe3\x83\xa0 (OS) \xe3\x81\xa7\xe3\x81\x99\xe3\x80\x82\xe3\x82\xaa\xe3\x83\x9a\xe3\x83\xac\xe3\x83\xbc\xe3\x83\x86\xe3\x82\xa3\xe3\x83\xb3\xe3\x82\xb0\xe3\x82\xb7\xe3\x82\xb9\xe3\x83\x86\xe3\x83\xa0\xe3\x81\xaf\xe3\x80\x81\xe3\x82\xb3\xe3\x83\xb3\xe3\x83\x94\xe3\x83\xa5\xe3\x83\xbc\xe3\x82\xbf\xe3\x82\x92\xe5\x8b\x95\xe4\xbd\x9c\xe3\x81\x95\xe3\x81\x9b\xe3\x82\x8b\xe3\x81\x9f\xe3\x82\x81\xe3\x81\xab\xe5\xbf\x85\xe8\xa6\x81\xe3\x81\xaa\xe5\x9f\xba\xe6\x9c\xac\xe3\x83\x97\xe3\x83\xad\xe3\x82\xb0\xe3\x83\xa9\xe3\x83\xa0\xe3\x81\xa8\xe3\x83\xa6\xe3\x83\xbc\xe3\x83\x86\xe3\x82\xa3\xe3\x83\xaa\xe3\x83\x86\xe3\x82\xa3\xe3\x81\xae\xe9\x9b\x86\xe5\x90\x88\xe4\xbd\x93\xe3\x81\xa7\xe3\x81\x99\xe3\x80\x82
Korean - \xeb\x8d\xb0\xeb\xb9\x84\xec\x95\x88\xec\x9d\x80 \xec\x97\xac\xeb\x9f\xac\xeb\xb6\x84\xec\x9d\x98 \xec\xbb\xb4\xed\x93\xa8\xed\x84\xb0\xeb\xa5\xbc \xec\x9c\x84\xed\x95\x9c \xec\x9e\x90\xec\x9c\xa0 \xec\x9a\xb4\xec\x98\x81\xec\xb2\xb4\xec\xa0\x9c\xec\x9e\x85\xeb\x8b\x88\xeb\x8b\xa4. \xec\x9a\xb4\xec\x98\x81\xec\xb2\xb4\xec\xa0\x9c\xeb\x9e\x80 \xec\xbb\xb4\xed\x93\xa8\xed\x84\xb0\xeb\xa5\xbc \xeb\x8f\x99\xec\x9e\x91\xec\x8b\x9c\xed\x82\xa4\xeb\x8a\x94 \xea\xb8\xb0\xeb\xb3\xb8\xec\xa0\x81\xec\x9d\xb8 \xed\x94\x84\xeb\xa1\x9c\xea\xb7\xb8\xeb\x9e\xa8\xea\xb3\xbc \xec\x9c\xa0\xed\x8b\xb8\xeb\xa6\xac\xed\x8b\xb0\xeb\xa5\xbc \xed\x95\x98\xeb\x82\x98\xeb\xa1\x9c \xeb\xac\xb6\xec\x9d\x80 \xea\xb2\x83\xec\x9e\x85\xeb\x8b\x88\xeb\x8b\xa4.
Russian - Debian -- \xd1\x8d\xd1\x82\xd0\xbe \xd1\x81\xd0\xb2\xd0\xbe\xd0\xb1\xd0\xbe\xd0\xb4\xd0\xbd\xd0\xb0\xd1\x8f \xd0\xbe\xd0\xbf\xd0\xb5\xd1\x80\xd0\xb0\xd1\x86\xd0\xb8\xd0\xbe\xd0\xbd\xd0\xbd\xd0\xb0\xd1\x8f \xd1\x81\xd0\xb8\xd1\x81\xd1\x82\xd0\xb5\xd0\xbc\xd0\xb0 (\xd0\x9e\xd0\xa1) \xd0\xb8 \xd0\xbd\xd0\xb0\xd0\xb1\xd0\xbe\xd1\x80 \xd0\xbf\xd1\x80\xd0\xb8\xd0\xba\xd0\xbb\xd0\xb0\xd0\xb4\xd0\xbd\xd1\x8b\xd1\x85 \xd0\xbf\xd1\x80\xd0\xbe\xd0\xb3\xd1\x80\xd0\xb0\xd0\xbc\xd0\xbc \xd0\xb4\xd0\xbb\xd1\x8f \xd0\xb2\xd0\xb0\xd1\x88\xd0\xb5\xd0\xb3\xd0\xbe \xd0\xba\xd0\xbe\xd0\xbc\xd0\xbf\xd1\x8c\xd1\x8e\xd1\x82\xd0\xb5\xd1\x80\xd0\xb0.
simplified Chinese - Debian \xe6\x98\xaf\xe4\xb8\x80\xe5\xa5\x97\xe4\xb8\xba\xe4\xba\x86\xe6\x82\xa8\xe7\x9a\x84\xe8\xae\xa1\xe7\xae\x97\xe6\x9c\xba\xe6\x89\x80\xe8\xae\xbe\xe8\xae\xa1\xe7\x9a\x84\xe8\x87\xaa\xe7\x94\xb1\xe6\x93\x8d\xe4\xbd\x9c\xe7\xb3\xbb\xe7\xbb\x9f (OS)\xe3\x80\x82\xe6\x93\x8d\xe4\xbd\x9c\xe7\xb3\xbb\xe7\xbb\x9f\xe6\x98\xaf\xe4\xbd\xbf\xe8\xae\xa1\xe7\xae\x97\xe6\x9c\xba\xe8\xbf\x90\xe8\xa1\x8c\xe7\x9a\x84\xe5\x9f\xba\xe6\x9c\xac\xe7\xa8\x8b\xe5\xba\x8f\xe5\x92\x8c\xe5\xb7\xa5\xe5\x85\xb7\xe7\x9a\x84\xe9\x9b\x86\xe5\x90\x88\xe3\x80\x82
EOF
} elsif ($locale eq "EUC-JP") {
    $str = <<"EOF";
Japanese - Debian \xa4\xcf\xa4\xa2\xa4\xca\xa4\xbf\xa4\xce\xa5\xb3\xa5\xf3\xa5\xd4\xa5\xe5\xa1\xbc\xa5\xbf\xa4\xce\xa4\xbf\xa4\xe1\xa4\xce\xa5\xd5\xa5\xea\xa1\xbc\xa4\xca\xa5\xaa\xa5\xda\xa5\xec\xa1\xbc\xa5\xc6\xa5\xa3\xa5\xf3\xa5\xb0\xa5\xb7\xa5\xb9\xa5\xc6\xa5\xe0 (OS) \xa4\xc7\xa4\xb9\xa1\xa3\xa5\xaa\xa5\xda\xa5\xec\xa1\xbc\xa5\xc6\xa5\xa3\xa5\xf3\xa5\xb0\xa5\xb7\xa5\xb9\xa5\xc6\xa5\xe0\xa4\xcf\xa1\xa2\xa5\xb3\xa5\xf3\xa5\xd4\xa5\xe5\xa1\xbc\xa5\xbf\xa4\xf2\xc6\xb0\xba\xee\xa4\xb5\xa4\xbb\xa4\xeb\xa4\xbf\xa4\xe1\xa4\xcb\xc9\xac\xcd\xd7\xa4\xca\xb4\xf0\xcb\xdc\xa5\xd7\xa5\xed\xa5\xb0\xa5\xe9\xa5\xe0\xa4\xc8\xa5\xe6\xa1\xbc\xa5\xc6\xa5\xa3\xa5\xea\xa5\xc6\xa5\xa3\xa4\xce\xbd\xb8\xb9\xe7\xc2\xce\xa4\xc7\xa4\xb9\xa1\xa3
EOF
} elsif ($locale eq "KOI8-R") {
    $str = <<"EOF";
Russian - Debian -- \xdc\xd4\xcf \xd3\xd7\xcf\xc2\xcf\xc4\xce\xc1\xd1 \xcf\xd0\xc5\xd2\xc1\xc3\xc9\xcf\xce\xce\xc1\xd1 \xd3\xc9\xd3\xd4\xc5\xcd\xc1 (\xef\xf3) \xc9 \xce\xc1\xc2\xcf\xd2 \xd0\xd2\xc9\xcb\xcc\xc1\xc4\xce\xd9\xc8 \xd0\xd2\xcf\xc7\xd2\xc1\xcd\xcd \xc4\xcc\xd1 \xd7\xc1\xdb\xc5\xc7\xcf \xcb\xcf\xcd\xd0\xd8\xc0\xd4\xc5\xd2\xc1.
EOF
} else {
    $str = <<"EOF";
English - Debian is a free operating system (OS) for your computer. An operating system is the set of basic programs and utilities that make your computer run.
EOF
}

chomp($str);

for($i=40; $i<80; $i++) {
	print "\x1b[2J\x1b[Hencoding=$locale,width=$i\n", "-" x $i , "\n";
	system("echo \"$str\" | dotextwrap $i");
	print "[ Hit Enter key. ]";
	<>;
}

for($i=40; $i<80; $i++) {
	print "\x1b[2J\x1b[Hencoding=$locale,width=$i\n", "-" x $i , "\n";
	system("echo \"$str\" | dotextwrap $i 8 \"/* \" \" * \"");
	print "[ Hit Enter key. ]";
	<>;
}

